rxandroidble/src/main/java/com/polidea/rxandroidble/internal/operations/CharacteristicLongWriteOperation.java

```diff
@@ -209,4 +213,48 @@ public Boolean call(Object emission) {
             }
         };
     }
+
+    private static Func1<Observable<? extends Throwable>, Observable<?>> retryOperationStrategy(
+            final WriteOperationRetryStrategy writeOperationRetryStrategy,
+            final ByteBuffer byteBuffer,
+            final int batchSize) {
+        return new Func1<Observable<? extends Throwable>, Observable<?>>() {
+            @Override
+            public Observable<?> call(Observable<? extends Throwable> observable) {
+                return observable
+                        .flatMap(applyRetryStrategy())
+                        .map(replaceByteBufferPositionForRetry());
+            }
+
+            @NonNull
+            private Func1<Throwable, Observable<WriteOperationRetryStrategy.LongWriteFailure>> applyRetryStrategy() {
+                return new Func1<Throwable, Observable<WriteOperationRetryStrategy.LongWriteFailure>>() {
+                    @Override
+                    public Observable<WriteOperationRetryStrategy.LongWriteFailure> call(Throwable cause) {
+                        if (!(cause instanceof BleException)) {
+                            return Observable.error(cause);
+                        }
+
+                        final int failedBatchNumber = (byteBuffer.position() - batchSize) / batchSize;
+                        final WriteOperationRetryStrategy.LongWriteFailure longWriteFailure =
+                                new WriteOperationRetryStrategy.LongWriteFailure(failedBatchNumber, (BleException) cause);
+
+                        return writeOperationRetryStrategy.call(Observable.just(longWriteFailure));
```

Comment:
Not an ideal solution since the `writeOperationRetryStrategy` will get called each time the a new failure will happen where is could be called only once per subscription. Checkout how the `Observable.compose()` function or `bufferIsNotEmptyAndOperationHasBeenAcknowledgedAndNotUnsubscribed` works.

Basically this: 
```
final int failedBatchNumber = (byteBuffer.position() - batchSize) / batchSize;
                         final WriteOperationRetryStrategy.LongWriteFailure longWriteFailure =
                                 new WriteOperationRetryStrategy.LongWriteFailure(failedBatchNumber, (BleException) cause);
```
could be separated to a `Observable.map()` and the `writeOperationRetryStrategy` could be treated as a `Observable.Transformer` (see `.compose()`)

Reply:
How close am I with something like that :) :
```java
            @Override
            public Observable<?> call(Observable<? extends Throwable> emittedOnWriteFailure) {
                return writeOperationRetryStrategy.call(
                        emittedOnWriteFailure.map(mapToLongWriteFailureObject())
                )
                        .doOnNext(replaceByteBufferPositionForRetry());
            }
```
No other `Throwable` than `BleException` can be emitted?

Reply:
No other than `BleGattCharacteristicException` can be retried... Again, new commit in this PR makes it easier to see the whole picture. :)

Reply:
Sorry, I've pushed :)

Reply:
Yup, I have seen—just have a bit of work now. Will get back to you as soon as possible. 👍

Reply:
I have checked out your branch and I hope I will have a moment Tomorrow to play with it and see if I can make it a little bit nicer. I cannot think of how to explain this so I will try to show it ;)